language/stdlib/modules/transaction_fee_distribution.mvir

```diff
@@ -0,0 +1,100 @@
+//
+// Current limitations are:
+// * We don't pay out per-epoch, and instead pay-out per-block.
+// * In the handling of dust, it could be beneficial to be the first validator in the validating
+//   set. However, the maximal benefit that can be achieved per-block(/epoch) is at most 1 microlibra.
+module TransactionFeeDistribution {
+  import 0x0.LibraCoin;
+  import 0x0.Block;
+  import 0x0.ValidatorSet;
+  import 0x0.LibraAccount;
+  import 0x0.TransactionFeeCollection;
+
+   public distribute_transaction_fees() {
+     let num_validators: u64;
+     let amount_to_distribute_per_validator: u64;
+     let amount_remainder: u64;
+     let lib_coin_to_distribute: LibraCoin.T;
+     let amount_collected: u64;
+
+     lib_coin_to_distribute, amount_collected = TransactionFeeCollection.relinquish_transaction_fees();
+     num_validators = ValidatorSet.number_of_validators();
+
+     // Calculate the amount of money to be dispursed, along with the remainder.
+     amount_to_distribute_per_validator, amount_remainder = Self.calculate_remainder(
+         copy(amount_collected),
+         move(num_validators)
+     );
+ 
+     // Iterate through the validators distributing fees equally
+     Self.distribute_transaction_fees_internal(
+         move(lib_coin_to_distribute),
+         move(amount_to_distribute_per_validator),
+         move(amount_remainder)
+     );
+ 
+     return;
+   }
+ 
+   distribute_transaction_fees_internal(
+       collected_fees: LibraCoin.T,
+       amount_to_distribute_per_validator: u64,
+       amount_remainder: u64
+   ) {
+       let index: u64;
+       let addr: address;
+       let keep_iter: bool;
+       let payment_amount: u64;
+       let payment_coin: LibraCoin.T;
+ 
+       index = 0;
+       keep_iter = false;
+       payment_amount = 0;
+ 
+       // Get the current address and whether this is a valid address for a validator.
+       addr, keep_iter = ValidatorSet.iter_validator_addresses(copy(index));
+ 
+       while (copy(keep_iter)) {
+ 
+           payment_amount = copy(amount_to_distribute_per_validator);
+ 
+           // If there is dust remaining, then pay to the validator.
+           if (copy(amount_remainder) > 0) {
+               payment_amount = move(payment_amount) + 1;
+               // decrement the amount of gas remaining.
+               amount_remainder = move(amount_remainder) - 1;
+           }
+ 
+           collected_fees, payment_coin = LibraCoin.split(
+               move(collected_fees),
+               copy(payment_amount)
+           );
+ 
+           LibraAccount.deposit(copy(addr), move(payment_coin));
+ 
+           // Increment the index into the validator set.
+           index = copy(index) + 1;
+ 
+           // Get the current address and whether we should keep iterating.
+           addr, keep_iter = ValidatorSet.iter_validator_addresses(copy(index));
+       }
+ 
+       LibraCoin.destroy_zero(move(collected_fees));
```

Comment:
I disagree that the fact that `amount_remainder` is always zero is nonobvious, but I agree that it may be hard for a verifier to prove. 

I  think paying it to the first validator might be a better solution if we decide not to go the way that it's currently implemented (since then it's paid to a validator that was part of the epoch, as opposed to one that may not have been a part of that epoch).